akka-http/src/main/scala/akka/http/scaladsl/server/directives/FramedEntityStreamingDirectives.scala

```diff
@@ -0,0 +1,141 @@
+    streamInternal[T](framing, (ec, mat) ⇒ Flow[ByteString].mapAsyncUnordered(parallelism)(Unmarshal(_).to[T](um, ec, mat)))
+  final def streamAsyncUnordered[T](parallelism: Int, framing: FramingWithContentType)(implicit um: Unmarshaller[ByteString, T]): RequestToSourceUnmarshaller[T] =
+    streamAsyncUnordered(parallelism)(um, framing)
+
+  // TODO materialized value may want to be "drain/cancel" or something like it?
+  // TODO could expose `streamMat`, for more fine grained picking of Marshaller
+
+  // format: OFF
+  private def streamInternal[T](framing: FramingWithContentType, marshalling: (ExecutionContext, Materializer) => Flow[ByteString, ByteString, NotUsed]#ReprMat[T, NotUsed]): RequestToSourceUnmarshaller[T] =
+    Unmarshaller.withMaterializer[HttpRequest, Source[T, NotUsed]] { implicit ec ⇒ implicit mat ⇒ req ⇒
+      val entity = req.entity
+      if (!framing.supported(entity.contentType)) {
+        val supportedContentTypes = framing.supported.map(ContentTypeRange(_))
+        FastFuture.failed(Unmarshaller.UnsupportedContentTypeException(supportedContentTypes))
+      } else {
+        val stream = entity.dataBytes.via(framing.flow).via(marshalling(ec, mat)).mapMaterializedValue(_ => NotUsed)  
+        FastFuture.successful(stream)
+      }
+    }
+  // format: ON
+
+  // TODO note to self - we need the same of ease of streaming stuff for the client side - i.e. the twitter firehose case.
+
+  implicit def _sourceMarshaller[T, M](implicit m: ToEntityMarshaller[T], mode: SourceRenderingMode): ToResponseMarshaller[Source[T, M]] =
+    Marshaller[Source[T, M], HttpResponse] { implicit ec ⇒ source ⇒
+      FastFuture successful {
+        Marshalling.WithFixedContentType(mode.contentType, () ⇒ { // TODO charset?
+          val bytes = source
+            .mapAsync(1)(t ⇒ Marshal(t).to[HttpEntity])
+            .map(_.dataBytes)
+            .flatMapConcat(ConstantFun.scalaIdentityFunction)
+            .intersperse(mode.start, mode.between, mode.end)
+          HttpResponse(entity = HttpEntity(mode.contentType, bytes))
+        }) :: Nil
+      }
+    }
+
+  implicit def _sourceParallelismMarshaller[T](implicit m: ToEntityMarshaller[T], mode: SourceRenderingMode): ToResponseMarshaller[AsyncRenderingOf[T]] =
+    Marshaller[AsyncRenderingOf[T], HttpResponse] { implicit ec ⇒ rendering ⇒
+      FastFuture successful {
+        Marshalling.WithFixedContentType(mode.contentType, () ⇒ { // TODO charset?
+          val bytes = rendering.source
+            .mapAsync(rendering.parallelism)(t ⇒ Marshal(t).to[HttpEntity])
+            .map(_.dataBytes)
+            .flatMapConcat(ConstantFun.scalaIdentityFunction)
+            .intersperse(mode.start, mode.between, mode.end)
+          HttpResponse(entity = HttpEntity(mode.contentType, bytes))
+        }) :: Nil
+      }
+    }
+
+  implicit def _sourceUnorderedMarshaller[T](implicit m: ToEntityMarshaller[T], mode: SourceRenderingMode): ToResponseMarshaller[AsyncUnorderedRenderingOf[T]] =
```

Comment:
Yes, it's a pattern applied in the `Directives` to make them not pop up in auto completion as much as the ones we want people to call.